thriftbp/ttl_client.go

```diff
@@ -17,50 +22,135 @@ const DefaultMaxConnectionAgeJitter = 0.1
 
 var _ Client = (*ttlClient)(nil)
 
+type ttlClientLockedData struct {
+	client     thrift.TClient
+	transport  thrift.TTransport
+	expiration time.Time // if expiration is zero, then the client will be kept open indefinetly.
+	timer      *time.Timer
+	closed     bool
+}
+
 // ttlClient is a Client implementation wrapping thrift's TTransport with a TTL.
 type ttlClient struct {
-	thrift.TClient
+	// configs needed for refresh to work
+	generator ttlClientGenerator
+	ttl       time.Duration
 
-	transport thrift.TTransport
+	replaceCounter metrics.Counter
 
-	// if expiration is zero, then the client will be kept open indefinetly.
-	expiration time.Time
+	// data guarded by lock (buffer-1 channel)
+	locked chan *ttlClientLockedData
 }
 
 // Close implements Client interface.
 //
 // It calls underlying TTransport's Close function.
 func (c *ttlClient) Close() error {
-	return c.transport.Close()
+	data := <-c.locked
+	defer func() {
+		c.locked <- data
+	}()
```

Comment:
🔕 gofmt will let you have this as a one-liner

Reply:
no it doesn't :) (we have gofmt in CI and it didn't complain)